src/field_theory/adjoin.lean

```diff
@@ -0,0 +1,204 @@
+def adjoin_as_submodule : submodule F E := {
+  carrier := adjoin F S,
+  zero_mem' := is_add_submonoid.zero_mem,
+  add_mem' := λ a b, is_add_submonoid.add_mem,
+  smul_mem' :=
+  begin
+    intros a b hb,
+    rw algebra.smul_def,
+    exact is_submonoid.mul_mem (adjoin.field_mem F S a) hb,
+  end
+}
+
+/-- proves an obvious linear equivalence (occasionally useful when working with dimension) -/
+definition adjoin_as_submodule_equiv : (adjoin F S) ≃ₗ[F] (adjoin_as_submodule F S) := {
+  to_fun := λ x, x,
+  map_add' := λ x y, rfl,
+  map_smul' :=
+  begin
+    intros x y,
+    ext1,
+    change _ = x • ↑y,
+    rw algebra.smul_def,
+    rw algebra.smul_def,
+    refl,
+  end,
+  inv_fun := λ x, x,
+  left_inv := λ x, rfl,
+  right_inv := λ x, rfl,
+}
+
+/-- If F ⊆ T and S ⊆ T then F[S] ⊆ F[T] -/
+lemma adjoin_subset {T : set E} [is_subfield T] (HF : set.range (algebra_map F E) ⊆ T)
+(HS : S ⊆ T) : adjoin F S ⊆ T :=
+begin
+  apply field.closure_subset,
+  rw set.union_subset_iff,
+  exact ⟨HF,HS⟩,
+end
+
+/-- If S ⊆ F[T] then F[S] ⊆ F[T] -/
+lemma adjoin_subset' {T : set E} (HT : S ⊆ adjoin F T) : adjoin F S ⊆ adjoin F T :=
+adjoin_subset F S (adjoin.field_subset F T) HT
+
+lemma set_range_subset {T₁ T₂ : set E} [is_subfield T₁] {hyp : T₁ ⊆ T₂} :
+set.range (algebra_map T₁ E) ⊆ T₂ :=
+begin
+  intros x hx,
+  cases hx with f hf,
+  rw ←hf,
+  cases f with t ht,
+  exact hyp ht,
+end
+
+lemma adjoin_contains_field_subset {F : set E} {HF : is_subfield F} {T : set E} {HT : T ⊆ F} :
+T ⊆ adjoin F S :=
+λ x hx, adjoin.field_mem F S ⟨x,HT hx⟩
+
+/-- F[S][T] = F[S ∪ T] -/
+lemma adjoin_twice (T : set E) : adjoin (adjoin F S) T = adjoin F (S ∪ T) :=
+begin
+  apply set.eq_of_subset_of_subset,
+  apply adjoin_subset,
+  apply set_range_subset,
+  apply adjoin_subset,
+  apply adjoin.field_subset,
+  apply adjoin_contains_subset,
+  apply set.subset_union_left,
+  apply adjoin_contains_subset,
+  apply set.subset_union_right,
+  apply adjoin_subset,
+  transitivity adjoin F S,
+  apply adjoin.field_subset,
+  apply adjoin_subset,
+  apply adjoin_contains_field_subset,
+  apply adjoin.field_subset,
+  apply adjoin_contains_field_subset,
+  apply adjoin.set_subset,
+  apply set.union_subset,
+  apply adjoin_contains_field_subset,
+  apply adjoin.set_subset,
+  apply adjoin.set_subset,
+end
+
+lemma adjoin.composition :
+(algebra_map F E) = (algebra_map (adjoin F S) E).comp (algebra_map F (adjoin F S)) := by ext;refl
+
+instance adjoin_algebra_tower : is_scalar_tower F (adjoin F S) E := {
+  smul_assoc :=
+  begin
+    intros x y z,
+    rw algebra.smul_def,
+    rw algebra.smul_def,
+    rw algebra.smul_def,
+    rw ring_hom.map_mul,
+    rw mul_assoc,
+    refl,
+  end
+}
+
+variables (α : E)
+
+--we're working on automatically getting notation for K[α,β,γ], but haven't quite figured it out yet
+notation K`[`:std.prec.max_plus β`]` := adjoin K (@singleton _ _ set.has_singleton β)
+notation K`[`:std.prec.max_plus β `,` γ`]` := adjoin K {β,γ}
```

Comment:
The following almost works:

```suggestion
notation K`[`:std.prec.max_plus l:(foldr `, ` (h t, set.insert h t) ∅) `]` := adjoin K l
```

Unfortunately, it will parse `F[α]` as `adjoin F (insert α ∅)` instead of `adjoin F (singleton α)`. This makes the proofs that follow slightly broken (`simp` should fix it), but the generality might be worth the slight awkwardness.